src/common/control.c

```diff
@@ -0,0 +1,231 @@
+                    g_variant_new("(sis)", g.progress.target ?: "", g.progress.progress,
+                        g.progress.phase ?: ""));
+            return;
+        }
+    }
+
+    g_dbus_method_invocation_return_dbus_error(invocation, SFMF_DBUS_INTERFACE ".MethodCallError",
+            "Invalid method call");
+}
+
+static const GDBusInterfaceVTable sfmf_control_vtable = {
+    sfmf_control_method_call_cb,
+    NULL,
+    NULL,
+};
+
+void sfmf_control_init(struct SFMF_Control_Callbacks *callbacks, void *user_data)
+{
+    g.callbacks = callbacks;
+    g.callbacks_user_data = user_data;
+
+    g.own_name = g_bus_own_name(G_BUS_TYPE_SYSTEM, SFMF_DBUS_NAME, G_BUS_NAME_OWNER_FLAGS_NONE,
+            sfmf_control_bus_acquired_cb, sfmf_control_name_acquired_cb, sfmf_control_name_lost_cb,
+            &g, NULL);
+
+    while (!g.name_acquired) {
+        sfmf_control_process();
+    }
+
+    GError *error = NULL;
+    GDBusNodeInfo *node_info = g_dbus_node_info_new_for_xml(""
+        "<!DOCTYPE node PUBLIC \"-//freedesktop//DTD D-BUS Object Introspection 1.0//EN\"\n"
+        "  \"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd\">\n"
+        "<node name=\"" SFMF_DBUS_PATH "\">\n"
+        "  <interface name=\"" SFMF_DBUS_INTERFACE "\">\n"
+        "    <method name=\"Abort\">\n"
+        "      <arg name=\"result\" type=\"b\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <method name=\"GetProgress\">\n"
+        "      <arg name=\"subvolume\" type=\"s\" direction=\"out\" />\n"
+        "      <arg name=\"progress\" type=\"i\" direction=\"out\" />\n"
+        "      <arg name=\"phase\" type=\"s\" direction=\"out\" />\n"
+        "    </method>\n"
+        "    <signal name=\"Progress\">\n"
+        "      <arg name=\"subvolume\" type=\"s\" />\n"
+        "      <arg name=\"progress\" type=\"i\" />\n"
+        "      <arg name=\"phase\" type=\"s\" />\n"
+        "    </signal>\n"
+        "  </interface>\n"
+        "</node>\n"
+    "", &error);
+    if (!node_info) {
+        SFMF_FAIL("Could not compile D-Bus XML: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    g.object_registration = g_dbus_connection_register_object(g.connection, "/",
+            g_dbus_node_info_lookup_interface(node_info, SFMF_DBUS_INTERFACE),
+            &sfmf_control_vtable, &g, NULL, &error);
+
+    if (!g.object_registration) {
+        SFMF_FAIL("Could not register object on D-Bus: %s\n", error->message);
+        g_error_free(error);
+    }
+
+    g_dbus_node_info_unref(node_info);
+}
+
+void sfmf_control_process()
+{
+    while (g_main_context_iteration(NULL, FALSE)) {
+        // pump the main loop
+    }
+}
+
+void sfmf_control_set_progress(const char *target, int progress, const char *phase)
+{
+    GError *error = NULL;
+
+    if (g.progress.target) {
+        g_free(g.progress.target);
+    }
+    if (target) {
+        g.progress.target = g_strdup(target);
+    }
```

Comment:
Indeed! I'll change that to `g_free(g.progress.target), g.progress.target = NULL;` (similar to how you do it in mce, to "group" the free and the assignment).